registry/storage/driver/gcs/gcs.go

```diff
@@ -80,25 +83,42 @@ type driver struct {
 // Required parameters:
 // - bucket
 func FromParameters(parameters map[string]interface{}) (storagedriver.StorageDriver, error) {
-
 	bucket, ok := parameters["bucket"]
 	if !ok || fmt.Sprint(bucket) == "" {
 		return nil, fmt.Errorf("No bucket parameter provided")
 	}
 
-	keyfile, ok := parameters["keyfile"]
-	if !ok {
-		keyfile = ""
-	}
-
 	rootDirectory, ok := parameters["rootdirectory"]
 	if !ok {
 		rootDirectory = ""
 	}
+
+	var ts oauth2.TokenSource
+	var jwtConf *jwt.Config
+	if keyfile, ok := parameters["keyfile"]; ok {
+		jsonKey, err := ioutil.ReadFile(fmt.Sprint(keyfile))
+		if err != nil {
+			return nil, err
+		}
+		jwtConf, err = google.JWTConfigFromJSON(jsonKey, storage.ScopeFullControl)
+		if err != nil {
+			return nil, err
+		}
+		ts = jwtConf.TokenSource(context.Background())
+	} else {
+		var err error
+		ts, err = google.DefaultTokenSource(context.Background(), storage.ScopeFullControl)
+		if err != nil {
+			return nil, err
+		}
+	}
+
 	params := driverParameters{
-		fmt.Sprint(bucket),
-		fmt.Sprint(keyfile),
-		fmt.Sprint(rootDirectory),
+		bucket:        fmt.Sprint(bucket),
+		rootDirectory: fmt.Sprint(rootDirectory),
+		email:         jwtConf.Email,
```

Comment:
It looks like this could be a nil pointer dereference if `keyfile` is not provided